htmresearch/algorithms/column_pooler.py

```diff
@@ -174,16 +192,127 @@ def compute(self, feedforwardInput=(), lateralInputs=(),
+        predictedActiveInput = numpy.intersect1d(feedforwardInput, predictedInput)
+        predictedGrowthCandidates = numpy.intersect1d(feedforwardGrowthCandidates, predictedInput)
+        self._computeInferenceMode(predictedActiveInput, lateralInputs)
+        self._computeOnlineLearningMode(predictedActiveInput, lateralInputs,
+                                        feedforwardGrowthCandidates)
+      elif (numpy.abs(len(self.activeCells) - self.sdrSize)
+          > (self.learningTolerance*self.sdrSize)):
+        # If the pooler doesn't have a single representation, try to infer one,
+        # before actually attempting to learn.
+        self._computeInferenceMode(feedforwardInput, lateralInputs)
+        self._computeOnlineLearningMode(feedforwardInput, lateralInputs,
+                                        feedforwardGrowthCandidates)
+      else:
+        # If there isn't predicted input and we have a single SDR,
+        # we are extending that representation and should just learn.
+        self._computeOnlineLearningMode(feedforwardInput, lateralInputs,
+                                  feedforwardGrowthCandidates)
+        self._computeInferenceMode(feedforwardInput, lateralInputs)
+
+
+  def _computeOnlineLearningMode(self, feedforwardInput, lateralInputs,
+                                 feedforwardGrowthCandidates):
+    """
+    Learning mode: we are learning a new object in an online fashion. If there
+    is no prior activity, we randomly activate 'sdrSize' cells and create
+    connections to incoming input. If there was prior activity, we maintain it.
+    If we have a union, we simply do not learn at all.
+
+    These cells will represent the object and learn distal connections to each
+    other and to lateral cortical columns.
+
+    Parameters:
+    ----------------------------
+    @param  feedforwardInput (sequence)
+            Sorted indices of active feedforward input bits
+
+    @param  lateralInputs (list of sequences)
+            For each lateral layer, a list of sorted indices of active lateral
+            input bits
+
+    @param  feedforwardGrowthCandidates (sequence or None)
+            Sorted indices of feedforward input bits that the active cells may
+            grow new synapses to.  This is assumed to be the predicted active
+            cells of the input layer.
+    """
+    prevActiveCells = self.activeCells
+    cellsToLearn = set(self.activeCells)
+
+    # If there are not enough previously active cells, select random subset of
+    # cells to learn on.  If there were only some active cells, we assume that
+    # we are learning an object related to what is represented by those cells;
+    # if there were none, we are learning a new object.
+    # This case is the only way different object representations are created.
+    if len(cellsToLearn) < self.sdrSize:
+      if len(cellsToLearn) < self.sdrSize*(1 - self.learningTolerance):
```

Comment:
This "learning tolerance" is necessary because of the possibility of errors, basically.  It's not guaranteed that all cells will become active when they should, and it is quite possible for an extra cell or two to become active.  If we insisted on having _exactly_ the right number of cells active, we wipe the SDR far more often than is desirable, and end up relearning things that should have been grouped together.  This tends to lead to results like this, where the redundant representations harm performance noticeably:
![prediction performance with swapped stimuli no shared range](https://user-images.githubusercontent.com/9668988/28792141-74550ecc-75e3-11e7-84e8-d97dc7134048.png)